### PR TITLE
New ToggleButtons

### LINKED ISSUE #17830: Enhanced toggle button(radio button)
The reason for changing the toggle button is that we have multiple variants and to make it consistent with new forms design. Should also be used more in forms replacing drop-downs with few items.  

Currently we have three version that are should be replaced with one: 
Screenshot 2019-06-28 at 15 13 08
Screenshot 2019-06-28 at 15 12 57
Screenshot 2019-04-25 at 15 33 33

- Only one item can be chosen. 
- Only first and last item should have border radius(2px). First item on left side and last item on right side.  
- Keyboard control: Tab to toggle button and use arrow keys and enter to choose like the dropdown menu.
- Can have label if used in forms

Screenshot 2019-09-03 at 14 41 21

Either creating new component from scratch or use one of the old ones to build upon is fine.

### PR DESCRIPTION
**Pull request by tskarhed**
Closes #17830

Please try how the focus state is handled. Not sure how to make it good when it is focus and selected/not. The colors might also be a little bit off.

### CODE REVIEW

**Comment by torkelo** [comment]
I think 3 sizes would be nice. for the logs component, a more slim version would be useful. 
https://ant.design/components/radio/

**Comment by dehrax** [comment]
Here are the button sizes 
Screenshot from 2019-07-08 11-30-20

**Comment by dprokop** [comment]
image

The hover state seems strange to me with only text being highlighted. Maybe we could try with some very gentle background color, similar to focus state?

**Comment by dprokop** [comment]
output

I have also noticed ^^ size glitch when toggling

File: packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        });
+      })}
+    </div>
+  );
+};
+
+ToggleButtonGroup.displayName = 'ToggleButtonGroup';
+
+export { ToggleButtonGroup };
+export class ToggleButton extends PureComponent<ToggleButtonProps, ToggleButtonState> {
```
Comment (dprokop) [inline packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx]: Does ToggleButton follow the same sizing as regular buttons(I think it should)? Then we could use `AbstractButton` component for rendering ToggleButton

File: packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    } else {
+      return button;
+    }
+  }
+}
+
+const getButtonSize = (size: ButtonSize = 'md') => {
+  switch (size) {
+    case 'lg':
+      return { padding: 20, height: 40 };
```
Comment (dprokop) [inline packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx]: The sizings here should be based on theme. If not possible then we should choose ones that are the closest.

File: packages/grafana-ui/src/components/ToggleButtonGroup/types.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { ReactNode } from 'react';
+import { ButtonSize } from '../Button/AbstractButton';
+
+export interface ToggleButtonState {
```
Comment (dprokop) [inline packages/grafana-ui/src/components/ToggleButtonGroup/types.ts]: State and prop types should be in the same file as the component itself

File: packages/grafana-ui/src/components/index.ts [hunk 4]
```diff
@@ -14,6 +14,8 @@ export { IndicatorsContainer } from './Select/IndicatorsContainer';
 export { NoOptionsMessage } from './Select/NoOptionsMessage';
 export { default as resetSelectStyles } from './Select/resetSelectStyles';
 export { ButtonSelect } from './Select/ButtonSelect';
+export { ToggleButtonGroup, ToggleButton } from './ToggleButtonGroup/ToggleButtonGroup';
+export * from './ToggleButtonGroup/types';
```
Comment (dprokop) [inline packages/grafana-ui/src/components/index.ts]: Why are you exposing the types from grafana/ui?

File: packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+      return { padding: 16, height: 36 };
+  }
+};
+
+const getStyles = (theme: GrafanaTheme, buttonSize?: ButtonSize) => {
+  const borderRadius = 2;
+  const bg = theme.colors.bodyBg;
+  const activeBg = increaseContrast(theme.colors.bodyBg, 5);
+  const activeText = theme.colors.blueLight;
+  const borderColor = theme.isDark ? theme.colors.gray1 : theme.colors.gray2;
```
Comment (dprokop) [inline packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx]: should use `selectThemeVariant`

File: packages/grafana-ui/src/themes/dark.ts [hunk 6]
```diff
@@ -41,6 +41,7 @@ const basicColors = {
 const darkTheme: GrafanaTheme = {
   ...defaultTheme,
   type: GrafanaThemeType.Dark,
+  isDark: true,
```
Comment (dprokop) [inline packages/grafana-ui/src/themes/dark.ts]: This information is already available via theme.type, isn't it?

File: packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx (hunk 1 above)
Comment (dprokop) [inline packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx]: Please move the ToggleButton component to a separate file

File: packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    this.setState({
+      untouched: true,
+    });
+  }
+
+  componentDidUpdate(prevProps: ToggleButtonProps) {
+    //Override internal select toggle if props are updated
+    if (prevProps.selected !== this.props.selected) {
+      this.setState({
+        selected: this.props.selected,
```
Comment (dprokop) [inline packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx]: This is going to cause additional re-render. You can use https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops

**Comment by dehrax** [comment]
> output
> 
> I have also noticed ^^ size glitch when toggling

Ah, it was a thing with the font weight. The width changes with the font width. Seems like I forgot to remove it from the focus state. But the bigger question is whether or not we want a heavier font weight when focus/hover/active. The thicker the font, the more weird it will look on state change.

For the focus state I would prefer to remove the blue border, as the border would signify if it is selected or not. The way it is solved now is that focus styles are disabled when the button is being toggled.
 Basically:
hover and focus: only blue tinted background
toggle on/off: change color of border

File: packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+  }
+
+  render() {
+    const { tooltip } = this.props;
+
+    const button = <ThemeContext.Consumer>{theme => this.renderButton(theme)}</ThemeContext.Consumer>;
+
+    if (tooltip) {
+      return (
+        <Tooltip content={tooltip} placement="bottom">
```
Comment (dprokop) [inline packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx]: Ideally the tooltip should be implemented in an AbstractButton component so that we don't need to think about handling tooltip in multiple button implementations.

File: packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+  return (
+    <div>
+      <span className={styles.label}>{label}</span>
+      {React.Children.map(children, (child: any, index) => {
+        // Set correct style depending on first/middle/last
+        const style = index === 0 ? styles.first : index === children.length - 1 ? styles.last : styles.middle;
+
+        return React.cloneElement(child, {
+          className: style,
+          propsPriority,
```
Comment (dprokop) [inline packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx]: This seems suspicious... ;) Can you elaborate?

**Review by dprokop — CHANGES_REQUESTED** [review]

File: packages/grafana-ui/src/components/index.ts (hunk 4 above)
Comment (dehrax) [reply packages/grafana-ui/src/components/index.ts]: As the state is being passed into `onChange`. Could always just pass on just `selected` and `value` and just forget about the type alltogether.

https://github.com/grafana/grafana/pull/17871/files/fd7a43782e6dc117f2f4b0f26bc2af6a16391d14#diff-41a0c07d4ace9d7b762563c546be3f89R116

File: packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx (hunk 7 above)
Comment (dehrax) [reply packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx]: Did this because I didn't want to "pollute" the state since it is passed onto `onChange`, but maybe that's an irrational fear? `getDerivedStateFromProps` works for me.

File: packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx (hunk 9 above)
Comment (dehrax) [reply packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx]: It is used to set styles for the first and last child. Should probably set it with CSS `:first-child` and `:last-child` instead, but I wrote it before I had properly figured out emotion.

`propsPriority` is used to turn off the internal `selected` state in `ToggleButton`. This sets the prop for all buttons in the group.

File: packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx (hunk 1 above)
Comment (dehrax) [reply packages/grafana-ui/src/components/ToggleButtonGroup/ToggleButtonGroup.tsx]: If you ask me, I think the sizing for regular buttons are varying too much. Wouldn't make sense if `ToggleButton` is meant to be used as a form component.
Screenshot from 2019-07-08 15-35-17

**Comment by torkelo** [comment]
Do we need this many sizes? 3 sized seem like a good amount. I think the normal Button sizes also are too many :) 

We need one that is 32px high to match the navbar buttons. since in Explore the Logs|Metrics picker is used in the nav.

**Comment by torkelo** [comment]
Then design / look of the button still needs work, so maybe have to wait until Patrick's back

**Comment by kaydelaney** [comment]
I think maybe the semantics of the markup would be improved by using checkboxes and labels rather than buttons and spans. Maybe something to this effect: https://codepen.io/kaydelaney/pen/JgeqbW

File: packages/grafana-ui/src/components/ToggleButton/ToggleButton.tsx [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    return null;
+  }
+
+  renderButton(theme: GrafanaTheme) {
+    const styles = getStyles(theme);
+    // @ts-ignore
+    const { children, className, tooltip, controlled, ...otherProps } = this.props;
+    const activeStyle = this.state.selected ? styles.active : null;
+    return (
+      <Button
```
Comment (torkelo) [inline packages/grafana-ui/src/components/ToggleButton/ToggleButton.tsx]: wonder if it would be better to use radio inputs restyled instead, like in https://codepen.io/kaydelaney/pen/QWLyNRz

not sure what the benefit would be

Comment (dehrax) [reply packages/grafana-ui/src/components/ToggleButton/ToggleButton.tsx]: It feels cleaner and more semantically correct as Kay said. Should probably rename the component to reflect the radio functionality in that case, `RadioButton` and `RadioButtonGroup`

Comment (kaydelaney) [reply packages/grafana-ui/src/components/ToggleButton/ToggleButton.tsx]: Some benefits would be better semantics which is helpful for accessibility, as well as automatic keyboard control

Comment (torkelo) [reply packages/grafana-ui/src/components/ToggleButton/ToggleButton.tsx]: ok, lets do that, and rename it to RadioButton & RadioButtonGroup
